mypy/build.py

```diff
@@ -998,6 +999,117 @@ def generate_deps_for_cache(manager: BuildManager,
+    if not new_baseline:
+        if file.exists():
+            file.unlink()
+            print("No errors, baseline file removed")
+        else:
+            print("No errors, no baseline to write")
+        # Indicate that writing was canceled
+        manager.options.write_baseline = False
+        return
+    if new_baseline == manager.errors.original_baseline:
+        return
+    if not file.parent.exists():
+        file.parent.mkdir(parents=True)
+    data: Mapping[str, object]
+    if (
+        manager.options.baseline_format == "1.3"
+        or manager.options.baseline_format == "default"
+    ):
+        data = {
+            "files": new_baseline,
+            "format": "1.3",
+            "sources": [source.path for source in sources],
+        }
+    elif manager.options.baseline_format == "1.2":
+        data = {
+            "__baseline_metadata__": {
+                "format": "1.2", "sources": [source.path for source in sources]
+            },
+            **new_baseline,
+        }
+    else:
+        raise Exception(
+            f"Invalid/Unsupported baseline file format {manager.options.baseline_format}"
+        )
+    json.dump(
+        data,
+        file.open("w"),
+        indent=2,
+        sort_keys=True,
+    )
+
+
+def load_baseline(options: Options, errors: Errors, stdout: TextIO) -> None:
+    if not options.baseline_file:
+        return
+
+    from mypy import main
+
+    formatter = util.FancyFormatter(stdout, stderr, options.show_error_codes)
+    file = Path(options.baseline_file)
+
+    if not file.exists():
+        if options.baseline_file != defaults.BASELINE_FILE:
+            msg = formatter.style(
+                f"error: Baseline file {options.baseline_file!r} not found", 'red',
+                bold=True
+            )
+            main.fail(msg, stderr, options)
+        return
+    try:
+        data: Dict[str, object] = json.load(file.open())
+    except JSONDecodeError:
+        msg = formatter.style(
+            f"error: Baseline file {options.baseline_file!r} contains invalid JSON",
+            'red', bold=True
+        )
+        main.fail(msg, stderr, options)
+    # try to detect format:
+    format_ = data.get("format")
+    if not format_:
+        metadata: Optional[Dict[str, object]] = data.get("__baseline_metadata__")
+        format_ = metadata and metadata.get("format")
+    if not format_:
+        format_ = "1.2"
+    if not options.write_baseline and options.baseline_format == "default":
+        options.baseline_format = format_
+    if format_ == "1.3":
+        baseline_errors = data.get("files")
+        sources = data.get("sources")
+    elif format_ == "1.2":
+        metadata = data.pop("__baseline_metadata__", None)
+        sources = metadata and metadata.get("sources") or ["None"]
+        baseline_errors = data
+    else:
+        raise Exception(f"Invalid/Unsupported baseline file format: {options.baseline_format}")
```

Comment:
this error text is duplicated [here](https://github.com/KotlinIsland/basedmypy/pull/239/files#diff-ea1209369af1f6691068ac958d46d42ed5ab3a6d1529f63d891ca9cab58e3998R1046), maybe make a class for it like `UnsupportedBaselineException`